src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,319 @@
+    optional = parser.add_argument_group("optional arguments")
+
+    # Required flag argument
+    required.add_argument(
+        "--workspace_id",
+        required=True,
+        type=int,
+        help="Workspace addressed by the permanent ID",
+    )
+    required.add_argument(
+        "--project_dir",
+        type=str,
+        help="Path points to Collections repository",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
+    )
+    optional.add_argument(
+        "--output_dir",
+        type=str,
+        help="Output directoy to save genome objects.",
+    )
+    optional.add_argument(
+        "--job_dir",
+        type=str,
+        help="Job directoy of source link",
+    )
+    optional.add_argument(
+        "--source", type=str, default="WS", help="Source of data (default: WS)"
+    )
+    optional.add_argument(
+        "--workers",
+        type=int,
+        default=5,
+        help="Number of workers for multiprocessing",
+    )
+    optional.add_argument(
+        "--overwrite", action="store_true", help="Overwrite existing files."
+    )
+    optional.add_argument(
+        "--batch_size", type=int, default=10000, help="Batch size of object id"
+    )
+
+    args = parser.parse_args()
+
+    (
+        workspace_id,
+        project_dir,
+        root_dir,
+        output_dir,
+        job_dir,
+        source,
+        workers,
+        overwrite,
+        batch_size,
+    ) = (
+        args.workspace_id,
+        args.project_dir,
+        args.root_dir,
+        args.output_dir,
+        args.job_dir,
+        args.source,
+        args.workers,
+        args.overwrite,
+        args.batch_size,
+    )
+
+    if source not in SOURCE:
+        raise ValueError(f"Unexpected source. Currently supported sources: {SOURCE}")
+
+    uid = loader_helper.get_id()
+    username = loader_helper.get_username()
+
+    job_dir = job_dir or _make_job_dir(project_dir, username)
+    output_dir = output_dir or _make_output_dir(
+        root_dir, loader_common_names.SOURCE_DATA_DIR, source
+    )
+
+    os.makedirs(output_dir, exist_ok=True)
+    os.makedirs(job_dir, exist_ok=True)
+
+    service = Service(uid, job_dir)
+    service.start()
+
+    conf = Conf(job_dir, output_dir, workers)
+
+    visited = set()
+    for upa in os.listdir(output_dir):
+        visited.add(upa)
+
+    for obj_info in list_objects(
+        workspace_id, conf, batch_size, FILTER_OBJECTS_NAME_BY
+    ):
+        upa = "{6}_{0}_{4}".format(*obj_info)
+        if upa in visited and not overwrite:
+            raise ValueError(
+                "{} is already in {}. Please add --overwrite flag to redownload".format(
+                    upa, output_dir
+                )
+            )
+        conf.queue.put([upa, obj_info])
+
+    for i in range(workers + 1):
+        conf.queue.put([None, None])
```

Comment:
It's a signal for the worker to shut down: https://github.com/kbase/collections/pull/225/files#diff-1ce3e6c7567701005d836ab7d2569c5938318aa586369a8c3224eb495406546aR173-R174

It looks to me like `queue.get()` blocks forever if the queue is empty:
https://docs.python.org/3/library/queue.html#queue.Queue.get

Reply:
Whether that signal is the right signal I'm not sure. There might be a better option

Reply:
I'd probably just use `None` rather than `[None, None]`

Reply:
it returns a list of 2 variables, right? I was thinking of something like
`task = conf.queue.get)
 if not task: break
upa, obj_info = task`
then we don't need to push [None, None].

Reply:
What Tian said, although we'll still need to push _something_ to signal that there's nothing else to process. `None` would work for this case

Reply:
> conf.queue.get return 2 variables: upa, obj_info

Also, to clarify, it returns 2 variables because a list of 2 items was pushed to the queue. `get()` only returns 1 variable, which in this case is a list

Reply:
Updated using None!

Reply:
:+1: